src/Microsoft.ML.HalLearners/SymSgdClassificationTrainer.cs

```diff
@@ -0,0 +1,850 @@
+                                int numFinalPassesToTrainThoroughly = 2;
+                                // We also do not want to learn for more passes than what the user asked
+                                int numPassesForThisBatch = Math.Min(numPassesForABatch, _args.NumberOfIterations - iter - numFinalPassesToTrainThoroughly);
+                                // If all of this leaves us with 0 passes, then set numPassesForThisBatch to 1
+                                numPassesForThisBatch = Math.Max(1, numPassesForThisBatch);
+                                state.PassIteration = iter;
+                                Native.LearnAll(inputDataManager, tuneLR, ref lr, l2Const, piw, weights.Values, ref bias, numFeatures,
+                                    numPassesForThisBatch, numThreads, tuneNumLocIter, ref numLocIter, _args.Tol, _args.Shuffle, shouldInitialize, stateGCHandle);
+                                shouldInitialize = false;
+
+                                // Check if we are done with going through the data
+                                if (inputDataManager.FinishedTheLoad)
+                                {
+                                    iter += numPassesForThisBatch;
+                                    // Check if more passes are left
+                                    if (iter < _args.NumberOfIterations)
+                                        inputDataManager.RestartLoading(_args.Shuffle, Host);
+                                }
+
+                                // If more passes are left, load as much as possible
+                                if (iter < _args.NumberOfIterations)
+                                    inputDataManager.LoadAsMuchAsPossible();
+                            }
+                        }
+
+                        // Maps back the dense features that are mislocated
+                        if (numThreads > 1)
+                            Native.MapBackWeightVector(weights.Values, stateGCHandle);
+                        Native.DeallocateSequentially(stateGCHandle);
+                    }
+                }
+            }
+            finally
+            {
+                if (stateGCHandle.IsAllocated)
+                    stateGCHandle.Free();
+            }
+            return CreatePredictor(weights, bias);
+        }
+
+        private void CheckLabel(RoleMappedData examples, out int weightSetCount)
+        {
+            examples.CheckBinaryLabel();
+            weightSetCount = 1;
+        }
+
+        private long AcceleratedMemoryBudgetBytes => _args.MemorySize * 1024 * 1024;
+        private long UsedMemory { get; set; }
+
+        private static unsafe class Native
+        {
+            internal const string DllName = "SymSgdNative";
+            internal const string MklDllName = "MklImports";
+
+            static Native()
+            {
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```

Comment:
It may be nice to comment why this is only needed on Windows, and what makes it work on the other platforms.